Monitor EC2 instance stats

thibautcornolti opened the pull request:
Add a way to monitor the EC2 instances and store the informations in ElasticSearch.

schubev commented:
Yeah, I used `Fraction` when building that so that we’d get a precise output. In truth we don’t care much about rounding errors so it’s fine to use floats instead.

thibautcornolti committed:
Add a way to monitor the ec2 instances in ES

TRAC-596 #comment Add a way to monitor the ec2 instances in ES

giubil commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+	defaultRegion      = "us-west-2"
+	elasticsearchIndex = "awshourlyinstanceusage"
+	elasticsearchType  = "ec2instance"
+)
```
Should this 3 const be here, or should they be in a conf file ?
Since this project is open source, people might need to change those, it would be easier to have them in a config file.

schubev replied:
`defaultRegion` definitely shouldn’t be a `const`. Also it already exists in the `config` module I believe. The other two are fine by me.

giubil commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+	if err != nil {
+		logger.Error("Get caller identity failed", err)
+		return ""
+	}
```
This logs an error but doesn't return one, which doesn't really make sense

schubev replied:
Indeed, the idiomatic way to do this in Go is to return `(string, error)` where the `string` should be disregarded if the `error` is not nil.

giubil commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+// getInstanceTag formats an array of *ec2.Tag to an array
+// of InstanceTag.
+func getInstanceTag(tags []*ec2.Tag) []InstanceTag {
+	res := make([]InstanceTag, 0, 5)
```
Why use `5` in the make if you append to it ?

schubev replied:
It’s a micro-optimization to prevent some allocations, but it’s very minimal and makes the code more complex. I don’t think it should be kept in that case.

giubil commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+	instances, err := svc.DescribeInstances(nil)
+	if err != nil {
+		logger.Error("Describe instances failed", err)
+		return
```
Again you log an error but you don't return one

giubil commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+	regions, err := svc.DescribeRegions(nil)
+	if err != nil {
+		logger.Error("Describe regions failed", err)
+		return []string{}
```
Again you log an error but don't return one

giubil commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+		logger.Error("Describe regions failed", err)
+		return []string{}
+	}
+	res := make([]string, 0, 10)
```
Why use `10` in the make if you append to it ?

giubil commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+
+// getPropertiesMapping uses the reflection of the struct InstanceInfo and
+// InstanceTag to get the mapping of the ElasticSearch documents.
+func getPropertiesMapping(field interface{}) map[string]interface{} {
```
A function that takes `field interface{}` and returns `map[string]interface{}` is really not explicit.
Either stop using `interface{}` or explain them in the documentation

schubev replied:
Preferably : don’t use them.

giubil commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+		},
+	}
+	if _, err := client.CreateIndex(elasticsearchIndex).BodyJson(mapping).Do(context.Background()); err != nil {
+		logger.Error("ElasticSearch create index failed", err)
```
Again you log an error but don't return one

giubil commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+		Dimensions: dimensions,
+	})
+	if err != nil {
+		logger.Error("Fetch CPU stats from CloudWatch failed", err)
```
Again you log an error but don't return one

giubil commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+// importInstancesToES imports an array of InstanceInfo in ElasticSearch.
+// It calls createIndexES if the index doesn't exist.
+func importInstancesToES(instances []InstanceInfo, logger jsonlog.Logger) {
+	// TODO: Get credentials
```
Why create a PR if you still have `TODO` comments ?

giubil commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+	// TODO: Get credentials
+	client, err := es.NewSignedElasticClient(config.EsUrl, credentials.NewSharedCredentials("", "default"))
+	if err != nil {
+		logger.Error("Get ElasticSearch client failed", err)
```
Again you log an error but don't return one

giubil commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+		return
+	}
+	if exists, err := client.IndexExists(elasticsearchIndex).Do(context.Background()); err != nil {
+		logger.Error("ElasticSearch check if index exists failed", err)
```
Again you log an error but don't return one

giubil commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+	}
+	for _, instance := range instances {
+		if res, err := client.Index().Index(elasticsearchIndex).Type(elasticsearchType).BodyJson(instance).Do(context.Background()); err != nil {
+			logger.Error("Instance put in ES failed", err)
```
Again you log an error but don't return one

giubil commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+		go fetchInstancesList(creds, region, account, logger, instanceInfoChan)
+		instanceInfoChans = append(instanceInfoChans, instanceInfoChan)
+	}
+	instances := make([]InstanceInfo, 0, 20)
```
Why use `20` in the make if you append to it ?

giubil commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+	)
+	account := getAccountID(creds, logger)
+	if err != nil {
+		logger.Error("Get temporary credentials failed", err)
```
Again you log an error but don't return one

giubil commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+		instances = append(instances, instance)
+	}
+	importInstancesToES(instances, logger)
+}
```
Please also write unit tests for your code

schubev commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+var (
+	// tagsAvailable is used to reflect the structs
+	// InstanceTag and InstanceInfo.
+	tagsAvailable = []string{"type", "index"}
```
Why use a slice instead of two constants ?

thibautcornolti replied:
It was to loop on it. In that way you could add tags without change the code (except this and the struct).
But it's removed, I changed the way.

schubev commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+	// InstanceInfo represents all the informations of an EC2 instance.
+	// It will be imported in ElasticSearch thanks to the struct tags.
+	InstanceInfo struct {
+		Account    string        `json:"account" type:"string" index:"not_analyzed"`
```
AFAIK type `string` is deprecated in ES 5.5, instead use type `keyword`. Please check about `not_analyzed`, too.

thibautcornolti replied:
Replaced by `keyword` and `text`. `keyword` is equal to `string not_analyzed` and `text` tokenize the field.

schubev commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+}
+
+// getAccountID allows to get the AWS Account ID (12 digits) of the
+// credentials given.
```
That’s french with english words. How about this ?
```go
// getAccountId gets the AWS Account ID for the given credentials
```

schubev replied:
Also, all other names we’ve written which contain acronyms treat them like normal words, so please name this `getAccountId`.

schubev commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+
+// createIndexES creates the ElasticSearch index with the properties
+// from getPropertiesMapping.
+func createIndexES(client *elastic.Client, logger jsonlog.Logger) {
```
The code shouldn’t create the mappings. Instead write a template to be provided to the ElasticSearch before Trackit is started.

schubev commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+
+// importInstancesToES imports an array of InstanceInfo in ElasticSearch.
+// It calls createIndexES if the index doesn't exist.
+func importInstancesToES(instances []InstanceInfo, logger jsonlog.Logger) {
```
Same thing as earlier about the acronyms in identifiers.

schubev commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+// It calls createIndexES if the index doesn't exist.
+func importInstancesToES(instances []InstanceInfo, logger jsonlog.Logger) {
+	// TODO: Get credentials
+	client, err := es.NewSignedElasticClient(config.EsUrl, credentials.NewSharedCredentials("", "default"))
```
You shouldn’t create your own `elastic.Client` when module `es` provides one already.

schubev replied:
Especially since this one is not properly configured : it will work neither in a development environment on a non-AWS elasticsearch, nor on AWS EC2 using EC2 roles.

schubev commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+		CPUPeak    float64       `json:"cpupeak" type:"double"`
+		KeyPair    string        `json:"keypair" type:"string" index:"not_analyzed"`
+		Type       string        `json:"type" type:"string" index:"not_analyzed"`
+		Tags       []InstanceTag `json:"tags" type:"nested"`
```
ElasticSearch documents representing AWS objects with tags should use the following scheme :
```json
{
    "tags": {
        "foo": "foovalue",
        "bar": "barvalue",
    }
}
```
What you are doing is this.
```json
{
    "tags": [
        {
            "key": "foo",
            "value": "foovalue"
        },
        {
            "key": "bar",
            "value": "barvalue"
        }
    ]
}
```

Use a `map[string]string` instead and ensure your mapping templates includes a dynamic mapping so these are of `keyword` type in ElasticSearch.

schubev commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+			Value: aws.String(instanceId),
+		},
+	}
+	stats, err := svc.GetMetricStatistics(&cloudwatch.GetMetricStatisticsInput{
```
Please use `context.Context` to
— provide cancelation of long, blocking operations like network communications
— log stuff (please log more)

schubev commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+// to import them in ElasticSearch. The stats are fetched from the last hour.
+// In this way, FetchInstancesStats should be called every hour.
+func FetchInstancesStats(awsAccount AwsAccount) {
+	logger := jsonlog.DefaultLogger
```
Please use a `context.Context` and get your logger from there.

schubev commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+func FetchInstancesStats(awsAccount AwsAccount) {
+	logger := jsonlog.DefaultLogger
+	tcreds, err := GetTemporaryCredentials(context.Background(), awsAccount, "fetchInstanceStats")
+	creds := credentials.NewStaticCredentials(
```
Using STS credentials this way means your function will start failing after one hour. Please look at function `getServiceForRepository` in file _/aws/s3/readBills.go_ for a correct example.

schubev commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+
+// fetchRegionsList fetchs the regions list from AWS and returns an array of their name.
+func fetchRegionsList(creds *credentials.Credentials, logger jsonlog.Logger) []string {
+	sess := session.Must(session.NewSession(&aws.Config{
```
`session.Session` is safe to use concurrently and it is better not to create one each time you need to use AWS. Please create your session and EC2 service in `FetchInstanceStats` and pass the service around instead of recreating it everywhere.

schubev commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+		return
+	}
+	now := time.Now()
+	hour := time.Date(now.Year(), now.Month(), now.Day(), now.Hour()-1, 0, 0, 0, now.Location())
```
Please extract that in a documented function.

schubev commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+// getInstanceStats gets the CPU average and the CPU peak from CloudWatch
+func getInstanceStats(instanceId string, sess *session.Session, region string, logger jsonlog.Logger) (float64, float64) {
+	now := time.Now()
+	end := time.Date(now.Year(), now.Month(), now.Day(), now.Hour()-1, 0, 0, 0, now.Location())
```
Please try and find a better way to do that, and otherwise extract it in its own documented function.

schubev commented on aws/monitorinstances.go:
```diff
@@ -0,0 +1,308 @@
+}
+
+// getInstanceStats gets the CPU average and the CPU peak from CloudWatch
+func getInstanceStats(instanceId string, sess *session.Session, region string, logger jsonlog.Logger) (float64, float64) {
```
Don’t pass a logger, pass a context.

thibautcornolti committed:
monitor ec2: Cleaned code and changed way to import stats

 - Rewrote comments
 - Added error handling
 - Added document versioning
 - Externalized ES mapping
 - Improved tags handling

TRAC-596

giubil commented on aws/ec2/monitorInstances.go:
```diff
@@ -0,0 +1,284 @@
+		MetricName: aws.String("CPUUtilization"),
+		StartTime:  aws.Time(start),
+		EndTime:    aws.Time(end),
+		Period:     aws.Int64(int64(86400 * 30)),
```
`86400 * 30` seems like a magic number. You should put it as a constant

giubil commented on aws/ec2/monitorInstances.go:
```diff
@@ -0,0 +1,284 @@
+	}
+}
+
+// fetchInstancesList, firstly, calls DescribeInstances from AWS to list all the instances in a region.
```
This type of comment does not really follow the coding style for Go comments. The first line should be a description of the function. Here the first line is not a complete description of what the function does

giubil commented on aws/ec2/monitorInstances.go:
```diff
@@ -0,0 +1,284 @@
+	instances, err := svc.DescribeInstances(nil)
+	if err != nil {
+		logger.Error("Error when describing instances", err.Error())
+		return
```
You log an error but don't return one.

giubil commented on aws/ec2/monitorInstances.go:
```diff
@@ -0,0 +1,284 @@
+	logger := jsonlog.LoggerFromContextOrDefault(ctx)
+	mapping := "{\"mappings\":{\"" + DocumentEc2 + "\":" + mappingEc2Instance + "}}"
+	if _, err := client.CreateIndex(IndexHourlyInstanceUsage).BodyString(mapping).Do(context.Background()); err != nil {
+		logger.Error("Error when creating ElasticSearch index", err.Error())
```
If your function can have an error, it should return one. How will the calling function know if this index creation has succeeded ? Should you continue the execution of your calling function if this function fails ?

giubil commented on aws/ec2/monitorInstances.go:
```diff
@@ -0,0 +1,284 @@
+		} else {
+			logger.Info("Instance put in ES", *res)
+		}
+	}
```
This function does again not return anything while function called inside of it can return errors.
Since you assume that this function will always succeed, what happens and it fails silently ? Will this be a problem for other functions depending on it ?

giubil commented on aws/ec2/mappings.go:
```diff
@@ -0,0 +1,61 @@
+	},
+	"numeric_detection": false,
+	"date_detection": false
+}`
```
The mapping is already being inserted inside of the ES via the script `./scripts/add_es_mapping.sh`. What is the purpose of this mapping ?

thibautcornolti replied:
It's used to automatically create the index if it doesn't exist. I was inspired by the work of @schubev on the PR#19

thibautcornolti committed:
monitor ec2: Changed comments, added error handling

TRAC-596
